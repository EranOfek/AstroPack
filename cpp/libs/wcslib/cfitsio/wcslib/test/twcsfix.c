/*============================================================================
  WCSLIB 7.3 - an implementation of the FITS WCS standard.
  Copyright (C) 1995-2020, Mark Calabretta

  This file is part of WCSLIB.

  WCSLIB is free software: you can redistribute it and/or modify it under the
  terms of the GNU Lesser General Public License as published by the Free
  Software Foundation, either version 3 of the License, or (at your option)
  any later version.

  WCSLIB is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
  more details.

  You should have received a copy of the GNU Lesser General Public License
  along with WCSLIB.  If not, see http://www.gnu.org/licenses.

  Direct correspondence concerning WCSLIB to mark@calabretta.id.au

  Author: Mark Calabretta, Australia Telescope National Facility, CSIRO.
  http://www.atnf.csiro.au/people/Mark.Calabretta
  $Id: twcsfix.c,v 7.3.1.2 2020/08/17 11:20:48 mcalabre Exp mcalabre $
*=============================================================================
*
* twcsfix tests the translation routines for non-standard WCS keyvalues, the
* wcsfix() suite, and the spectral coordinate translation routine wcssptr().
*
*---------------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <wcs.h>
#include <wcserr.h>
#include <wcsfix.h>
#include <wcsprintf.h>
#include <wcsunits.h>
#include <wcsutil.h>


void parser(struct wcsprm *);

const int NAXIS = 3;
const double CRPIX[3] =  {90.0, 90.0, 1.0};
const double PC[3][3] = {{ 1.0,  0.0, 0.0},
                         { 0.0,  1.0, 0.0},
                         { 0.0,  0.0, 1.0}};
const double CDELT[3] =  {-1.0, 1.0, 19.68717093222};

char CUNIT[3][9] = {"ARCSEC", "ARCSEC", "KM/SEC"};

// N.B. non-standard.
char CTYPE[3][9] = {"RA---NCP", "DEC--NCP", "FELO-HEL"};

const double CRVAL[3] = {265.6220947090*3600.0, -28.98849996030*3600.0,
                         5569.27104};
const double RESTFRQ = 1.42040575e9;
const double RESTWAV = 0.0;

// N.B. non-standard date-time format.
const char DATEOBS[] = "1957/02/15 01:10:00";
const char DATEBEG[] = "1957/02/15 01:10:00";
const char DATEAVG[] = "1957/02/15 02:10:00";
const char DATEEND[] = "1957/02/15 03:10:00";

const double BEPOCH  = 1957.124382563;
const double MJDBEG  = 35884.048611;

const double OBSGEO_L =   148.263510;
const double OBSGEO_B =   -32.998406;
const double OBSGEO_H =      411.793;

// For testing spcfix().
const int  VELREF = 2;
const char SPECSYS[] = "BARYCENT";

int main()

{
  char ctypeS[9];
  int i, stat[NWCSFIX], status;
  struct wcsprm wcs;
  struct wcserr info[NWCSFIX];

  wcsprintf("Testing WCSLIB translator for non-standard usage (twcsfix.c)\n"
          "------------------------------------------------------------\n\n");

  wcs.flag = -1;
  parser(&wcs);

  // Note: to print the unfixed wcsprm struct using wcsprt() the struct
  // would first have to be initialized by wcsset().  However, if the struct
  // contains non-standard keyvalues then wcsset() will either fix them
  // itself or else fail (e.g. for non-standard units).  Thus, in general,
  // wcsprt() cannot be used to print the unmodified struct.

  // Fix non-standard WCS keyvalues.
  wcserr_enable(1);
  status = wcsfixi(7, 0, &wcs, stat, info);
  wcsprintf("wcsfix status returns: (");
  for (i = 0; i < NWCSFIX; i++) {
    wcsprintf(i ? ", %d" : "%d", stat[i]);
  }
  wcsprintf(")\n");

  for (i = 0; i < NWCSFIX; i++) {
    if (info[i].status < -1 || 0 < info[i].status) {
      wcsprintf("\n");
      wcserr_prt(info+i, 0x0);

      // Free memory used to store the message.
      if (info[i].msg) wcsdealloc(info[i].msg);
    }
  }

  if (status) {
    wcsprintf("\nwcsfix error %d", status);
    return 1;
  }

  // Extract information from the FITS header.
  if (wcssetprm(&wcs)) {
    wcsprintf("\n");
    wcsperr(&wcs, 0x0);
  }

  wcsprintf("\n");
  wcsprt(&wcs);
  wcsprintf("\n------------------------------------"
            "------------------------------------\n");

  // Should now have a 'VOPT-F2W' axis, translate it to frequency.
  strcpy(ctypeS, "FREQ-???");
  i = -1;
  if (wcssptr(&wcs, &i, ctypeS)) {
    wcsperr(&wcs, 0x0);
    return 1;
  }

  if (wcssetprm(&wcs)) {
    wcsperr(&wcs, 0x0);
    return 1;
  }

  wcsprt(&wcs);

  wcsfree(&wcs);

  return 0;
}

//----------------------------------------------------------------------------

void parser(wcs)

struct wcsprm *wcs;

{
  int i, j;
  double *pcij;

  // In practice a parser would read the FITS header until it encountered
  // the NAXIS keyword which must occur near the start, before any of the
  // WCS keywords.  It would then use wcsini() to allocate memory for
  // arrays in the wcsprm struct and set default values.  In this
  // simulation the header keyvalues are set as global variables.
  wcsnpv(2);
  wcsini(1, NAXIS, wcs);


  // Now the parser scans the FITS header, identifying WCS keywords and
  // loading their values into the appropriate elements of the wcsprm
  // struct.

  for (j = 0; j < NAXIS; j++) {
    wcs->crpix[j] = CRPIX[j];
  }

  pcij = wcs->pc;
  for (i = 0; i < NAXIS; i++) {
    for (j = 0; j < NAXIS; j++) {
      *(pcij++) = PC[i][j];
    }
  }

  for (i = 0; i < NAXIS; i++) {
    wcs->cdelt[i] = CDELT[i];
  }

  for (i = 0; i < NAXIS; i++) {
    strcpy(wcs->cunit[i], &CUNIT[i][0]);
  }

  for (i = 0; i < NAXIS; i++) {
    strcpy(wcs->ctype[i], &CTYPE[i][0]);
  }

  for (i = 0; i < NAXIS; i++) {
    wcs->crval[i] = CRVAL[i];
  }

  wcs->restfrq = RESTFRQ;
  wcs->restwav = RESTWAV;

  wcs->pv[0].i = -1;
  wcs->pv[0].m = -1;
  wcs->pv[0].value = -1.0;
  wcs->npv = 1;

  wcs->velref = VELREF;

  // dateobs and datebeg will be set by datfix().
  strcpy(wcs->dateavg, DATEAVG);
  strcpy(wcs->dateend, DATEEND);
  wcs->bepoch = BEPOCH;
  wcs->mjdbeg = MJDBEG;

  wcs->obsgeo[3] = OBSGEO_L;
  wcs->obsgeo[4] = OBSGEO_B;
  wcs->obsgeo[5] = OBSGEO_H;

  strcpy(wcs->specsys, SPECSYS);

  return;
}
