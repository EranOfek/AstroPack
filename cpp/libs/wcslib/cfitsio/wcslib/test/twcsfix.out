Testing WCSLIB translator for non-standard usage (twcsfix.c)
------------------------------------------------------------

wcsfix status returns: (-1, 0, 0, 0, 0, 0, -1)

INFORMATIVE message from datfix() at line 589 of file wcsfix.c:
Set MJD-OBS to 35884.048611 from BEPOCH.
Set DATE-OBS to '1957-02-15T01:10:00.018' from MJD-OBS.
Set DATE-BEG to '1957-02-15T01:09:59.990' from MJD-BEG.
Set MJD-AVG to 35884.090278 from DATE-AVG.
Changed DATE-AVG from '1957/02/15 02:10:00' to '1957-02-15T02:10:00'.
Set MJD-END to 35884.131944 from DATE-END.
Changed DATE-END from '1957/02/15 03:10:00' to '1957-02-15T03:10:00'.

INFORMATIVE message from obsfix() at line 694 of file wcsfix.c:
Set OBSGEO-X to -4554231.538 from OBSGEO-[LBH].
Set OBSGEO-Y to  2816759.125 from OBSGEO-[LBH].
Set OBSGEO-Z to -3454036.303 from OBSGEO-[LBH].

INFORMATIVE message from unitfix() at line 834 of file wcsfix.c:
Changed units:
  'ARCSEC' -> 'arcsec',
  'ARCSEC' -> 'arcsec',
  'KM/SEC' -> 'km/s.

INFORMATIVE message from spcfix() at line 882 of file wcsfix.c:
Changed CTYPE3 from 'FELO-HEL' to 'VOPT-F2W' (VELREF=2).

       flag: 137
      naxis: 3
      crpix: 0x<address>
               90.000       90.000       1.0000    
         pc: 0x<address>
    pc[0][]:   1.0000       0.0000       0.0000    
    pc[1][]:   0.0000       1.0000       0.0000    
    pc[2][]:   0.0000       0.0000       1.0000    
      cdelt: 0x<address>
              -0.00027778   0.00027778   19687.    
      crval: 0x<address>
               265.62      -28.988       5.5693e+06
      cunit: 0x<address>
             "deg"
             "deg"
             "m/s"
      ctype: 0x<address>
             "RA---SIN"
             "DEC--SIN"
             "VOPT-F2W"
    lonpole: 180.000000
    latpole: -28.988500
    restfrq: 1420405750.000000
    restwav: 0.000000
        npv: 3
     npvmax: 3
         pv: 0x<address>
              -1  -1  -1.0000    
               2   1   0.0000    
               2   2  -1.8049    
        nps: 0
     npsmax: 8
         ps: 0x<address>
         cd: 0x<address>
    cd[0][]:   0.0000       0.0000       0.0000    
    cd[1][]:   0.0000       0.0000       0.0000    
    cd[2][]:   0.0000       0.0000       0.0000    
      crota: 0x<address>
               0.0000       0.0000       0.0000    
     altlin: 0
     velref: 2
        alt: ' '
     colnum: 0
      colax: 0x<address>
                 0      0      0
      cname: 0x<address>
             UNDEFINED
             UNDEFINED
             UNDEFINED
      crder: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED
      csyer: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED
      czphs: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED
      cperi: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED
    wcsname: UNDEFINED
    timesys: UNDEFINED
    trefpos: UNDEFINED
    trefdir: UNDEFINED
    plephem: UNDEFINED
   timeunit: UNDEFINED
    dateref: UNDEFINED
     mjdref:      0.000000000     0.000000000
   timeoffs: UNDEFINED
    dateobs: "1957-02-15T01:10:00.018"
    datebeg: "1957-02-15T01:09:59.990"
    dateavg: "1957-02-15T02:10:00"
    dateend: "1957-02-15T03:10:00"
     mjdobs:  35884.048611317
     mjdbeg:  35884.048611000
     mjdavg:  35884.090277778
     mjdend:  35884.131944444
     jepoch: UNDEFINED
     bepoch:   1957.124382563
     tstart: UNDEFINED
      tstop: UNDEFINED
    xposure: UNDEFINED
    telapse: UNDEFINED
    timsyer: UNDEFINED
    timrder: UNDEFINED
    timedel: UNDEFINED
   timepixr: UNDEFINED
     obsgeo:  -4554231.538450  2816759.125233 -3454036.303343
                   148.263510      -32.998406      411.793000
   obsorbit: UNDEFINED
    radesys: "ICRS"
    equinox: UNDEFINED
    specsys: "BARYCENT"
    ssysobs: UNDEFINED
    velosys: UNDEFINED
    zsource: UNDEFINED
    ssyssrc: UNDEFINED
    velangl: UNDEFINED
        aux: 0x0
       ntab: 0
        tab: 0x0
       nwtb: 0
        wtb: 0x0
      types: 0x<address>
            2200 2201 3300
     lngtyp: "RA"
     lattyp: "DEC"
        lng: 0
        lat: 1
       spec: 2
   cubeface: -1
        err: 0x0
        lin: (see below)
        cel: (see below)
        spc: (see below)
     m_flag: 137
    m_naxis: 3
    m_crpix: 0x<address>  (= crpix)
       m_pc: 0x<address>  (= pc)
    m_cdelt: 0x<address>  (= cdelt)
    m_crval: 0x<address>  (= crval)
    m_cunit: 0x<address>  (= cunit)
    m_ctype: 0x<address>  (= ctype)
       m_pv: 0x<address>  (= pv)
       m_ps: 0x<address>  (= ps)
       m_cd: 0x<address>  (= cd)
    m_crota: 0x<address>  (= crota)

    m_colax: 0x<address>  (= colax)
    m_cname: 0x<address>  (= cname)
    m_crder: 0x<address>  (= crder)
    m_csyer: 0x<address>  (= csyer)
    m_czphs: 0x<address>  (= czphs)
    m_cperi: 0x<address>  (= cperi)
      m_aux: 0x0  (= aux)
      m_tab: 0x0  (= tab)
      m_wtb: 0x0  (= wtb)

   lin.*
       flag: 137
      naxis: 3
      crpix: 0x<address>
               90.000       90.000       1.0000    
         pc: 0x<address>
    pc[0][]:   1.0000       0.0000       0.0000    
    pc[1][]:   0.0000       1.0000       0.0000    
    pc[2][]:   0.0000       0.0000       1.0000    
      cdelt: 0x<address>
              -0.00027778   0.00027778   19687.    
     dispre: 0x0
     disseq: 0x0
     piximg: (nil)
     imgpix: (nil)
    i_naxis: 0
      unity: 1
     affine: 1
     simple: 1
        err: 0x0
     tmpcrd: 0x<address>
     m_flag: 0
    m_naxis: 0
    m_crpix: 0x0
       m_pc: 0x0
    m_cdelt: 0x0
   m_dispre: 0x0
   m_disseq: 0x0

   cel.*
      flag: 137
     offset: 0
       phi0:  0.000000
     theta0: 90.000000
        ref:   265.62      -28.988       180.00      -28.988    
        prj: (see below)
      euler:   265.62       118.99       180.00      -0.48463      0.87472   
    latpreq: 0 (not required)
     isolat: 0
        err: 0x0

   prj.*
       flag: 105
       code: "SIN"
         r0: 57.295780
         pv: (0)
              0.0000      -1.8049    
       phi0:  0.000000
     theta0: 90.000000
     bounds: 7

       name: "orthographic/synthesis"
   category: 1 (zenithal)
    pvrange: 102
  simplezen: 0
  equiareal: 0
  conformal: 0
     global: 0
  divergent: 0
         x0: 0.000000
         y0: 0.000000
        err: 0x0
        w[]:   0.017453     3.2577       4.2577       2.2577       0.0000    
               0.0000       0.0000       0.0000       0.0000       0.0000    
          m: 0
          n: 0
     prjx2s: 0x<address>
     prjs2x: 0x<address>

   spc.*
       flag: 111
       type: "VOPT"
       code: "F2W"
      crval:  5.5693e+06
    restfrq: 1420405750.000000
    restwav: 0.000000
         pv: (not used)
          w:   0.21106      1.3945e+09  -4.5667      (remainder unused)
    isGrism: 0
        err: 0x0
     spxX2P: 0x<address>
     spxP2S: 0x<address>
     spxS2P: 0x<address>
     spxP2X: 0x<address>

------------------------------------------------------------------------
       flag: 137
      naxis: 3
      crpix: 0x<address>
               90.000       90.000       1.0000    
         pc: 0x<address>
    pc[0][]:   1.0000       0.0000       0.0000    
    pc[1][]:   0.0000       1.0000       0.0000    
    pc[2][]:   0.0000       0.0000       1.0000    
      cdelt: 0x<address>
              -0.00027778   0.00027778  -89906.    
      crval: 0x<address>
               265.62      -28.988       1.3945e+09
      cunit: 0x<address>
             "deg"
             "deg"
             "Hz"
      ctype: 0x<address>
             "RA---SIN"
             "DEC--SIN"
             "FREQ"
    lonpole: 180.000000
    latpole: -28.988500
    restfrq: 1420405750.000000
    restwav: 0.000000
        npv: 3
     npvmax: 3
         pv: 0x<address>
              -1  -1  -1.0000    
               2   1   0.0000    
               2   2  -1.8049    
        nps: 0
     npsmax: 8
         ps: 0x<address>
         cd: 0x<address>
    cd[0][]:   0.0000       0.0000       0.0000    
    cd[1][]:   0.0000       0.0000       0.0000    
    cd[2][]:   0.0000       0.0000       0.0000    
      crota: 0x<address>
               0.0000       0.0000       0.0000    
     altlin: 0
     velref: 2
        alt: ' '
     colnum: 0
      colax: 0x<address>
                 0      0      0
      cname: 0x<address>
             UNDEFINED
             UNDEFINED
             UNDEFINED
      crder: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED
      csyer: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED
      czphs: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED
      cperi: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED
    wcsname: UNDEFINED
    timesys: UNDEFINED
    trefpos: UNDEFINED
    trefdir: UNDEFINED
    plephem: UNDEFINED
   timeunit: UNDEFINED
    dateref: UNDEFINED
     mjdref:      0.000000000     0.000000000
   timeoffs: UNDEFINED
    dateobs: "1957-02-15T01:10:00.018"
    datebeg: "1957-02-15T01:09:59.990"
    dateavg: "1957-02-15T02:10:00"
    dateend: "1957-02-15T03:10:00"
     mjdobs:  35884.048611317
     mjdbeg:  35884.048611000
     mjdavg:  35884.090277778
     mjdend:  35884.131944444
     jepoch: UNDEFINED
     bepoch:   1957.124382563
     tstart: UNDEFINED
      tstop: UNDEFINED
    xposure: UNDEFINED
    telapse: UNDEFINED
    timsyer: UNDEFINED
    timrder: UNDEFINED
    timedel: UNDEFINED
   timepixr: UNDEFINED
     obsgeo:  -4554231.538450  2816759.125233 -3454036.303343
                   148.263510      -32.998406      411.793000
   obsorbit: UNDEFINED
    radesys: "ICRS"
    equinox: UNDEFINED
    specsys: "BARYCENT"
    ssysobs: UNDEFINED
    velosys: UNDEFINED
    zsource: UNDEFINED
    ssyssrc: UNDEFINED
    velangl: UNDEFINED
        aux: 0x0
       ntab: 0
        tab: 0x0
       nwtb: 0
        wtb: 0x0
      types: 0x<address>
            2200 2201 3000
     lngtyp: "RA"
     lattyp: "DEC"
        lng: 0
        lat: 1
       spec: 2
   cubeface: -1
        err: 0x0
        lin: (see below)
        cel: (see below)
        spc: (see below)
     m_flag: 137
    m_naxis: 3
    m_crpix: 0x<address>  (= crpix)
       m_pc: 0x<address>  (= pc)
    m_cdelt: 0x<address>  (= cdelt)
    m_crval: 0x<address>  (= crval)
    m_cunit: 0x<address>  (= cunit)
    m_ctype: 0x<address>  (= ctype)
       m_pv: 0x<address>  (= pv)
       m_ps: 0x<address>  (= ps)
       m_cd: 0x<address>  (= cd)
    m_crota: 0x<address>  (= crota)

    m_colax: 0x<address>  (= colax)
    m_cname: 0x<address>  (= cname)
    m_crder: 0x<address>  (= crder)
    m_csyer: 0x<address>  (= csyer)
    m_czphs: 0x<address>  (= czphs)
    m_cperi: 0x<address>  (= cperi)
      m_aux: 0x0  (= aux)
      m_tab: 0x0  (= tab)
      m_wtb: 0x0  (= wtb)

   lin.*
       flag: 137
      naxis: 3
      crpix: 0x<address>
               90.000       90.000       1.0000    
         pc: 0x<address>
    pc[0][]:   1.0000       0.0000       0.0000    
    pc[1][]:   0.0000       1.0000       0.0000    
    pc[2][]:   0.0000       0.0000       1.0000    
      cdelt: 0x<address>
              -0.00027778   0.00027778  -89906.    
     dispre: 0x0
     disseq: 0x0
     piximg: (nil)
     imgpix: (nil)
    i_naxis: 0
      unity: 1
     affine: 1
     simple: 1
        err: 0x0
     tmpcrd: 0x<address>
     m_flag: 0
    m_naxis: 0
    m_crpix: 0x0
       m_pc: 0x0
    m_cdelt: 0x0
   m_dispre: 0x0
   m_disseq: 0x0

   cel.*
      flag: 137
     offset: 0
       phi0:  0.000000
     theta0: 90.000000
        ref:   265.62      -28.988       180.00      -28.988    
        prj: (see below)
      euler:   265.62       118.99       180.00      -0.48463      0.87472   
    latpreq: 0 (not required)
     isolat: 0
        err: 0x0

   prj.*
       flag: 105
       code: "SIN"
         r0: 57.295780
         pv: (0)
              0.0000      -1.8049    
       phi0:  0.000000
     theta0: 90.000000
     bounds: 7

       name: "orthographic/synthesis"
   category: 1 (zenithal)
    pvrange: 102
  simplezen: 0
  equiareal: 0
  conformal: 0
     global: 0
  divergent: 0
         x0: 0.000000
         y0: 0.000000
        err: 0x0
        w[]:   0.017453     3.2577       4.2577       2.2577       0.0000    
               0.0000       0.0000       0.0000       0.0000       0.0000    
          m: 0
          n: 0
     prjx2s: 0x<address>
     prjs2x: 0x<address>

   spc.*
       flag: 0
       type: "    "
       code: "   "
      crval: UNDEFINED
    restfrq: 0.000000
    restwav: 0.000000
         pv: (not used)
          w:   0.0000       0.0000       0.0000      (remainder unused)
    isGrism: 0
        err: 0x0
     spxX2P: 0x0
     spxP2S: 0x0
     spxS2P: 0x0
     spxP2X: 0x0
